### PR TITLE
[wasm-metadata] pretty-print cli output

### LINKED ISSUE #1923: [wasm-metadata] redesign the human-readable terminal output
I've been having some trouble parsing the textual output of the `wasm-tools metadata show` command. Since the command supports a `--json` flag to make the output machine-interpretable, I wanted to float the idea of changing the default textual rendering.

## Inspiration

I'm not an expert on CLI interfaces, so I was asking around for good references of CLI tool output, and I was pointed towards the output of [muesli/duf](https://github.com/muesli/duf). This creates a table for each device category, being relatively easy to parse on a glance.

ascii tables in a terminal

## Current Output

Here is an example of what the current output looks like for a `wasi:http` component, written in Rust. If you don't understand the underlying component/provider data model, making sense of the output can be pretty difficult. In preparing to file this issue I've read the `@provider` spec in `webassembly/tool-conventions`, so I understand the hierarchy now - but I wish this would have been easier.

```text
component:
    processed-by:
        wit-component: 0.216.0
    module rust_wasi_hello.wasm:
        language:
            Rust
            C11
        processed-by:
            rustc: 1.82.0 (f6e511eec 2024-10-15)
            clang: 18.1.2-wasi-sdk (https://github.com/llvm/llvm-project 26a1d6601d727a96f4301d0d8647b5a42760ae0c)
            wit-component: 0.20.1
            wit-bindgen-rust: 0.33.0
            wit-bindgen-c: 0.17.0
    module wit-component:adapter:wasi_snapshot_preview1:
        language:
            Rust
        processed-by:
            rustc: 1.80.0 (051478957 2024-07-21)
    module wit-component:shim:
        processed-by:
            wit-component: 0.216.0
    module wit-component:fixups:
        processed-by:
            wit-component: 0.216.0
    component:
```

## Proposed Output

Here is a sketch I made using Monodraw of what `wasm-tools metadata show` output could look like if we took inspiration from `duf`. This creates metadata per-component and module, making it easier to visually track which parts of the binary carry which metadata.

The idea would be to also add some basic colors to make the output easier to scan too. But as I mentioned earlier on: we wouldn't be changing the `--json` output at all. That way this only affects the human-readable output, and not the machine-readable one.

```text
┌──────────────────────────────────────┐                                            
│ component sample-wasi-http-rust      │                                            
├──────────┬────────────────┬──────────┤                                            
│ KIND     │ PRODUCER       │ VERSION  │                                            
├──────────┼────────────────┼──────────┤                                            
│ tool     │ wit-component  │ 0.216.0  │                                            
└──────────┴────────────────┴──────────┘                                            
┌──────────────────────────────────────────────────────────────────────────────────┐
│ module sample_wasi_http_rust.wasm                                                │
├──────────┬──────────────────┬────────────────────────────────────────────────────┤
│ KIND     │ PRODUCER         │ VERSION                                            │
├──────────┼──────────────────┼────────────────────────────────────────────────────┤
│ language │ Rust             │ -                                                  │
├──────────┼──────────────────┼────────────────────────────────────────────────────┤
│ language │ C11              │ -                                                  │
├──────────┼──────────────────┼────────────────────────────────────────────────────┤
│ tool     │ rustc            │ 1.82.0 (f6e511eec 2024-10-15)                      │
├──────────┼──────────────────┼────────────────────────────────────────────────────┤
│ tool     │ clang            │ 18.1.2-wasi-sdk                                    │
│          │                  │ (https://github.com/llvm/llvm-project              │
├──────────┼──────────────────┼────────────────────────────────────────────────────┤
│ tool     │ wit-component    │ 0.20.1                                             │
├──────────┼──────────────────┼────────────────────────────────────────────────────┤
│ tool     │ wit-bindgen-rust │ 0.20.1                                             │
├──────────┼──────────────────┼────────────────────────────────────────────────────┤
│ tool     │ wit-bindgen-c    │ 0.20.1                                             │
└──────────┴──────────────────┴────────────────────────────────────────────────────┘
┌──────────────────────────────────────────────────────┐                            
│ module wit-component:adapter:wasi_snapshot_preview1: │                            
├──────────┬──────────┬────────────────────────────────┤                            
│ KIND     │ PRODUCER │ VERSION                        │                            
├──────────┼──────────┼────────────────────────────────┤                            
│ language │ Rust     │ -                              │                            
├──────────┼──────────┼────────────────────────────────┤                            
│ tool     │ rustc    │ 1.82.0 (f6e511eec 2024-10-15)  │                            
└──────────┴──────────┴────────────────────────────────┘                            
┌────────────────────────────────────┐                                              
│ module wit-component-shim          │                                              
├──────────┬───────────────┬─────────┤                                              
│ KIND     │ PRODUCER      │ VERSION │                                              
├──────────┼───────────────┼─────────┤                                              
│ language │ Rust          │ -       │                                              
├──────────┼───────────────┼─────────┤                                              
│ tool     │ wit-component │ 0.216.0 │                                              
└──────────┴───────────────┴─────────┘                                              
┌────────────────────────────────────┐                                              
│ module wit-component:fixups        │                                              
├──────────┬───────────────┬─────────┤                                              
│ KIND     │ PRODUCER      │ VERSION │                                              
├──────────┼───────────────┼─────────┤                                              
│ language │ Rust          │ -       │                                              
├──────────┼───────────────┼─────────┤                                              
│ tool     │ wit-component │ 0.216.0 │                                              
└──────────┴───────────────┴─────────┘

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Closes #1923. While this doesn't quite deliver on the vision I sketched in the issue, I believe it's a meaningful step up from the status quo. I've also filed https://github.com/Nukesor/comfy-table/issues/160, which if accepted will hopefully allow us to more closely match `duf`'s output.

I'm not quite sure what the process for introducing new dependencies here is. To pretty-print tables, we're going to need *some* printing library - and `comfy-table` seemed like a reasonable choice. I looked at the deps, and nothing particularly stood out to me as a potential problem?

Thanks!

## Screenshot
This is what the output currently looks like. I might do another pass to add colors, etc - but this seemed like a good starting point.

Screenshot 2024-12-10 at 03 08 38

### CODE REVIEW

**Comment by yoshuawuyts** [comment]
Ok, this should be good to review now!

File: crates/wasm-metadata/Cargo.toml [hunk 1]
```diff
@@ -13,6 +13,7 @@ workspace = true
 [dependencies]
 clap = { workspace = true, optional = true }
 anyhow = { workspace = true }
+comfy-table = { workspace = true }
```
Comment (alexcrichton) [inline crates/wasm-metadata/Cargo.toml:16]: Could this perhaps be deferred to the CLI? I could imagine a `--format json` flag eventually being added to the CLI so having this as a CLI-only dependency I think might be better to have

**Review by alexcrichton — COMMENTED** [review]
Looks reasonable to me! I think it's fine to pull in deps like this for the CLI yeah.

File: crates/wasm-metadata/Cargo.toml (hunk 1 above)
Comment (yoshuawuyts) [reply crates/wasm-metadata/Cargo.toml:16]: The CLI already has a working `--json` flag, here is what the output looks like for this branch:
Screenshot 2024-12-10 at 17 53 43

That's not to say I'm not happy to move it though - would you still prefer we move it?

Comment (alexcrichton) [reply crates/wasm-metadata/Cargo.toml:16]: Oh hey nice! Personally though I still would prefer to move this to the CLI as that's sort of where "rendering" happens and this is a render-only dependency which someone programmatically using `wasm-metadata` might not need.

**Comment by yoshuawuyts** [comment]
updated!

**Review by alexcrichton — APPROVED** [review]

**Pull request merged**